### ISSUE TITLE
Issue with running FirefoxDriver in threaded environment (Java)

### ISSUE THREAD

**Issue by dmand** [SeleniumHQ/selenium#2226]
OS:   Ubuntu Linux 14.04
Selenium Version:  2.53.0
Browser:  Firefox
Browser Version: 46.0.1

---------------

Consider the following program:
```java
package seleniumdemo;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.chrome.ChromeDriver;

import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.logging.Logger;

public class Runner {
    private static final Logger log = Logger.getLogger(Runner.class.getName());

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        ExecutorService executor = Executors.newFixedThreadPool(6);

        ArrayList<Future> futures = new ArrayList<>();

        long startTime = System.currentTimeMillis();
        log.info("Starting tasks");
        for(int i = 0; i < 100; i++) {
            Future future = executor.submit(new Runnable() {
                @Override
                public void run() {
                    WebDriver driver = new FirefoxDriver();
                    driver.get("http://blazedemo.com");
                    driver.findElement(By.cssSelector("input.btn.btn-primary")).click();
                    driver.findElement(By.xpath("//table[@class='table']/tbody/tr[4]/td[1]/input")).click();
                    driver.close();
                }
            });
            futures.add(future);
        }

        log.info("Awaiting tasks");
        for (Future future : futures)
            future.get();
        log.info("All tasks finished");

        long endTime = System.currentTimeMillis();
        double duration = (endTime - startTime) / 1000;
        log.info(String.format("Tests took %f seconds", duration));

        executor.shutdownNow();
    }

}
```

This simple program creates a thread pool of 6 threads and then attempts to execute 100 simple  WebDriver tasks in parallel, distributed between these threads. It takes ~4 minutes to complete when using FirefoxDriver (and i can obverse that at most two instances of Firefox are opened at the same time, most of the time it's only one instance). If i replace FirefoxDriver with ChromeDriver - this program takes only ~1 minute to complete (and i can observe all 6 browsers opened at the same time).

I don't think that this behaviour is expected, so it's probably a bug on FirefoxDriver side.

Here's the JVM thread dump in the middle of execution:
```
"pool-1-thread-6" prio=10 tid=0x00007f205c303000 nid=0x68a0 waiting for monitor entry [0x00007f205031a000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.openqa.selenium.internal.SocketLock.lock(SocketLock.java:81)
	- waiting to lock <0x00000007b1102948> (a java.lang.Object)
	at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.start(NewProfileExtensionConnection.java:80)
	at org.openqa.selenium.firefox.FirefoxDriver.startClient(FirefoxDriver.java:271)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:119)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:216)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:211)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:207)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:120)
	at seleniumdemo.Runner$1.run(Runner.java:29)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

"pool-1-thread-5" prio=10 tid=0x00007f205c301800 nid=0x689f waiting for monitor entry [0x00007f205041b000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.openqa.selenium.internal.SocketLock.lock(SocketLock.java:81)
	- waiting to lock <0x00000007b1102948> (a java.lang.Object)
	at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.start(NewProfileExtensionConnection.java:80)
	at org.openqa.selenium.firefox.FirefoxDriver.startClient(FirefoxDriver.java:271)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:119)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:216)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:211)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:207)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:120)
	at seleniumdemo.Runner$1.run(Runner.java:29)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

"pool-1-thread-4" prio=10 tid=0x00007f205c300000 nid=0x689e waiting for monitor entry [0x00007f205051c000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.openqa.selenium.internal.SocketLock.lock(SocketLock.java:81)
	- waiting to lock <0x00000007b1102948> (a java.lang.Object)
	at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.start(NewProfileExtensionConnection.java:80)
	at org.openqa.selenium.firefox.FirefoxDriver.startClient(FirefoxDriver.java:271)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:119)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:216)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:211)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:207)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:120)
	at seleniumdemo.Runner$1.run(Runner.java:29)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

"pool-1-thread-3" prio=10 tid=0x00007f205c2fe000 nid=0x689d waiting for monitor entry [0x00007f205061d000]
   java.lang.Thread.State: BLOCKED (on object monitor)
	at org.openqa.selenium.internal.SocketLock.lock(SocketLock.java:81)
	- waiting to lock <0x00000007b1102948> (a java.lang.Object)
	at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.start(NewProfileExtensionConnection.java:80)
	at org.openqa.selenium.firefox.FirefoxDriver.startClient(FirefoxDriver.java:271)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:119)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:216)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:211)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:207)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:120)
	at seleniumdemo.Runner$1.run(Runner.java:29)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

"Forwarding newSession on session null to remote" prio=10 tid=0x00007f205c2fc800 nid=0x689c runnable [0x00007f205071d000]
   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:152)
	at java.net.SocketInputStream.read(SocketInputStream.java:122)
	at org.apache.http.impl.io.SessionInputBufferImpl.streamRead(SessionInputBufferImpl.java:139)
	at org.apache.http.impl.io.SessionInputBufferImpl.fillBuffer(SessionInputBufferImpl.java:155)
	at org.apache.http.impl.io.SessionInputBufferImpl.readLine(SessionInputBufferImpl.java:284)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:140)
	at org.apache.http.impl.conn.DefaultHttpResponseParser.parseHead(DefaultHttpResponseParser.java:57)
	at org.apache.http.impl.io.AbstractMessageParser.parse(AbstractMessageParser.java:261)
	at org.apache.http.impl.DefaultBHttpClientConnection.receiveResponseHeader(DefaultBHttpClientConnection.java:165)
	at org.apache.http.impl.conn.CPoolProxy.receiveResponseHeader(CPoolProxy.java:167)
	at org.apache.http.protocol.HttpRequestExecutor.doReceiveResponse(HttpRequestExecutor.java:272)
	at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:124)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:271)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:71)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:55)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.fallBackExecute(ApacheHttpClient.java:144)
	at org.openqa.selenium.remote.internal.ApacheHttpClient.execute(ApacheHttpClient.java:90)
	at org.openqa.selenium.remote.HttpCommandExecutor.execute(HttpCommandExecutor.java:142)
	at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.execute(NewProfileExtensionConnection.java:160)
	at org.openqa.selenium.firefox.FirefoxDriver$LazyCommandExecutor.execute(FirefoxDriver.java:380)
	at org.openqa.selenium.remote.RemoteWebDriver.execute(RemoteWebDriver.java:644)
	at org.openqa.selenium.remote.RemoteWebDriver.startSession(RemoteWebDriver.java:249)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:131)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:216)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:211)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:207)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:120)
	at seleniumdemo.Runner$1.run(Runner.java:29)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)

"pool-1-thread-1" prio=10 tid=0x00007f205c2f2800 nid=0x689b waiting on condition [0x00007f205081f000]
   java.lang.Thread.State: TIMED_WAITING (sleeping)
	at java.lang.Thread.sleep(Native Method)
	at org.openqa.selenium.internal.SocketLock.lock(SocketLock.java:91)
	- locked <0x00000007b1102948> (a java.lang.Object)
	at org.openqa.selenium.firefox.internal.NewProfileExtensionConnection.start(NewProfileExtensionConnection.java:80)
	at org.openqa.selenium.firefox.FirefoxDriver.startClient(FirefoxDriver.java:271)
	at org.openqa.selenium.remote.RemoteWebDriver.<init>(RemoteWebDriver.java:119)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:216)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:211)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:207)
	at org.openqa.selenium.firefox.FirefoxDriver.<init>(FirefoxDriver.java:120)
	at seleniumdemo.Runner$1.run(Runner.java:29)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
```

From the looks of it, most of the threads are locked in `NewProfileExtensionConnection.start()` that uses `SocketLock`.

**Comment by AutomatedTester** [comment]
Unfortunately we don't make any claims to be threadsafe in the client code.

**Comment by dmand** [comment]
I see.

It would be nice to have a thread-safe FirefoxDriver though.

**Comment by nirvdrum** [comment]
Perhaps I missed something in the discussion, but it looks like it is thread-safe. The problem is it's not concurrent. Is that accurate?

**Comment by undera** [comment]
It's fine that you don't claim thread-safety. But it's very sad that you hurrying to close the request from the field. 

First of all, it IS thread-safe. The lock we see is the way to make it thread safe. But it is done by price of performance in concurrent case.

The question is only how can we make it to work fine in multi-threaded environment, without locking different drivers on each other for too long. It's not the locking that is problem, it's the time spent under the lock. Maybe there is a way to remove some lengthy operations from under the lock and that would solve the case.

**Comment by undera** [comment]
@nirvdrum, you're correct, the problem is that it spends too much time under the lock

**Comment by nirvdrum** [comment]
@undera I don't have the cycles to look that currently, but can you verify if you have the issue when using the Selenium Server to run your Firefox tests? This works a bit differently by way of the client speaking HTTP to a server process that communicates to the browser driver directly.

**Comment by dmand** [comment]
I've tried that.

I started `selenium-server-standalone-2.53.0.jar` and applied the following changes to my program:
```diff
-                    WebDriver driver = new FirefoxDriver();
+                    DesiredCapabilities capability = DesiredCapabilities.firefox();
+                    driver = new RemoteWebDriver(new URL("http://localhost:4444/wd/hub"),  capability);
```

The results are the same, only one (sometimes two) instances of Firefox are active at a time and running time is still around 4 minutes.
